src/main/java/com/squareup/javawriter/builders/TypeSpec.java

```diff
@@ -97,13 +108,48 @@ void emit(CodeWriter codeWriter, String enumName) {
       }
       codeWriter.emit(" {\n");
     } else if (anonymousTypeArguments != null) {
-      codeWriter.emit("new $T(", supertype);
+      codeWriter.emit("new $T(", getOnlyElement(superinterfaces, superclass));
       codeWriter.emit(anonymousTypeArguments);
       codeWriter.emit(") {\n");
     } else {
       codeWriter.emitAnnotations(annotations, false);
       codeWriter.emitModifiers(modifiers);
-      codeWriter.emit("$L $L {\n", Ascii.toLowerCase(type.name()), name.simpleName());
+      codeWriter.emit("$L $L", Ascii.toLowerCase(type.name()), name.simpleName());
+      codeWriter.emitTypeVariables(typeVariables);
+
+      List<TypeName> extendsTypes;
+      List<TypeName> implementsTypes;
+      if (type == Type.INTERFACE) {
+        extendsTypes = superinterfaces;
+        implementsTypes = ImmutableList.of();
+      } else {
+        extendsTypes = !superclass.equals(ClassName.OBJECT)
```

Comment:
I like to write ternarys as `(expression) ? (positive result) : (negative result)`. I think of having types to print is the positive result, and no types to print is the negative result.

It's annoying 'cause this expression is `(not degenerate)` but overall I think it's consistent.